pkg/git/git.go

```diff
@@ -0,0 +1,119 @@
+package git
+
+import (
+	"github.com/SAP/jenkins-library/pkg/log"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/transport/http"
+)
+
+// utilsWorkTree interface abstraction of git.Worktree to enable tests
+type utilsWorkTree interface {
+	Add(path string) (plumbing.Hash, error)
+	Commit(msg string, opts *git.CommitOptions) (plumbing.Hash, error)
+	Checkout(opts *git.CheckoutOptions) error
+}
+
+// utilsRepository interface abstraction of git.Repository to enable tests
+type utilsRepository interface {
+	Worktree() (*git.Worktree, error)
+	Push(o *git.PushOptions) error
+}
+
+// utilsGit interface abstraction of git to enable tests
+type utilsGit interface {
+	plainClone(path string, isBare bool, o *git.CloneOptions) (*git.Repository, error)
+}
+
+// CommitSingleFile Commits the file located in the relative file path with the commitMessage to the given worktree.
+// In case of errors, the error is returned. In the successful case the commit is provided.
+func CommitSingleFile(filePath, commitMessage string, worktree *git.Worktree) (plumbing.Hash, error) {
+	return commitSingleFile(filePath, commitMessage, worktree)
+}
+
+func commitSingleFile(filePath, commitMessage string, worktree utilsWorkTree) (plumbing.Hash, error) {
```

Comment:
I understand why you have to do this split into public and private function for PlainClone, because there you create the runtime instance of the interface yourself. But for the other public functions (CommitSingleFile, PushChangesToRepository and ChangeBranch), it should be enough to have only the public version which takes the interface parameter. The difference is that for plainClone you need to mock a function, whereas for the others, you want to be able to pass a different struct.

```suggestion
func CommitSingleFile(filePath, commitMessage string, worktree utilsWorkTree) (plumbing.Hash, error) {
```

Reply:
But wouldn't the consumers be confused when they are getting asked for a strange interface where they need to guess what kind of runtime object they could pass into?

Reply:
I'm torn. I really see your point, on the other hand interface parameters would be more flexible. Thanks for being so open to suggestions and input!